test/unit/math/mix/fun/simplex_constrain_test.cpp

```diff
@@ -23,6 +23,7 @@ void expect_simplex_transform(const T& x) {
   auto f2 = [](const auto& x) { return g2(x); };
   auto f3 = [](const auto& x) { return g3(x); };
   stan::test::expect_ad(f1, x);
+  stan::test::expect_ad_matvar(f1, x);
   stan::test::expect_ad(f2, x);
   stan::test::expect_ad(f3, x);
```

Comment:
Oh weird, the reverse mode specializations are missing for `f2` and `f3`? Make an issue. This is kindof weird and nobody will be back through here in a while so we should write it down even if it is a tiny thing.

I guess the larger issue could be making the `x_constrain(x, lp)` functions more efficient. For instance, in unit vector constrain there is an extra dot product that happens:

```
inline auto unit_vector_constrain(const T& y, var& lp) {
  auto x = unit_vector_constrain(y);
  lp -= 0.5 * dot_self(y);
  return x;
}
```

That `dot_self` is effectively already computed inside `unit_vector_constrain`.

Reply:
Yeah I would but tbh I'm just too dumb and couldn't figure them out lol. Would you mind adding these? Also where are these signatures used?

Reply:
`y = *_constrain(x)` goes from unconstrained to constrained space without Jacobian adjustment. This is probably not used much (actually optimization might use it).

`y = *_constrain(x, lp)` goes from unconstrained to constrained with Jacobian adjustment. It would be the one that is used for the constraints when you are sampling.

We should probably do some reworking to focus optimization efforts on the second.

If it's just these four constraints in here could probably rework this in a few hours of wiggling with the code. Want me to do it in this pull?

Reply:
@SteveBronder I looked into `simplex_constrain`. I have code for it, but there was some guess and check involved and I want to work it out properly to make sure I'm not wasting memory. Let's break out making the two argument constrains fast into another pull.

Reply:
(if this is okay, let me know and I'll make the issue and put the `simplex_constrain` code there until we have time to get around to it)